Don't include unit test sources in code coverage (RIPD-1132)

bachase opened the pull request:
Migrates all unit tests under `src/ripple` to `src/test`.  A unit test is any cpp that has the BEAST_DEFINE_TESTSUITE* macro in it.   This will exclude the unit test implementation from contributing to code coverage.

* Standardize unit test file names to `foo_test.cpp`
* A handful of tests were in the same cpp as the feature being tested, e.g. `Workers.cpp`.  I broke the test out to a standalone file in these cases.
* A handful of tests were only in the unity builds or excluded from all builds (at least under VS) and are now included in both unity and non-unity. 
* There were two `hash_speed_impl_test.cpp` files, but one was not included in any configuration and looked to be an older version of the one that was.  I deleted the old one.

Things I don't like:
* `src\unity\test_unity.cpp` is big and takes a long time to compile.
* I added snappy include directory for all unit test sources in scons and cmake, but only the nodestore tests need them.
* `beast_Zero_test.cpp` caused ADL issues in the unity build.  I changed a namespace to workaround and it is still testing ADL now, but someone with more ADL-fu may have a better solution.

bachase commented on CMakeLists.txt:
```diff
@@ -643,6 +639,12 @@ if (WIN32 OR is_xcode OR unity)
 
   list(APPEND src ${beast_unity_srcs} ${ripple_unity_srcs})
 
+  add_with_props(src/unity/test_unity.cpp
```
@seelabs this is some copy-pasta with the add_with_props below.  I wasn't sure the preferred way.

seelabs replied:
@bachase Yes, that's the right function to add a source that add properties just for that source.

seelabs replied:
@bachase Since both `test_unity.cpp` and `nodestore.cpp` use exactly the same properties, I would use the same `add_properties` call to add them.

The only way I know to do that is by assigning the two files to an intermediate variable and feeding that variable the the function (something like `add_with_props("${unity_nodestore_srcs}" `). There may be a better way to create the list inline, but I don't know it.

vinniefalco replied:
Mega ugly

vinniefalco replied:
Add "per-source-file-properties" to the document outlining our complications with the build system

bachase commented on src/test/beast_Zero_test.cpp:
```diff
@@ -31,7 +31,7 @@ int signum (adl_tester)
 
 
-namespace detail {
+namespace inner_adl_test {
```
Having the `signum` for `adl_tester` inside beast::detail was resulted in shadowing of the non-namespaced, generic `signum` in [Zero.h](https://github.com/ripple/rippled/blob/develop/src/ripple/beast/utility/Zero.h#L58)

vinniefalco commented:
ICE on rocksdb (Travis)

vinniefalco commented:
Yeah I agree, if `test_unity.cpp` is choking up builds then we should address. Its very easy, just split the file into N pieces (e.g. `test1_unity.cpp`, `test2_unity.cpp`, ...) and give each of them a roughly equal amount of work.

bachase commented:
I could also mirror the modules/directories in the `src/ripple` tree and have unity per each of those.

vinniefalco commented:
@bachase I'm not sure what the value is on echoing the layout of the modules. Its something we can always do later. And its likely that we will reshuffle files around (especially in src/ripple/app).

bachase commented:
@vinniefalco Since I'm new to code base it just helps me get my bearings. But if that structure is in flux, then clearly not worthwhile.

vinniefalco commented:
No Travis report?

bachase commented:
I forgot to rebase, but not sure why Travis is missing.  Will investigate some later today/tomorrow, but can close the PR if is too noisy.

vinniefalco commented:
No need to close the PR

bachase commented:
reviewers? @vinniefalco and @mellery451

vinniefalco commented:
I just had a really uhh... stupid/smart idea. We can put the test files in `src/test/ripple/test`. And then set the added C++ include path to `src/test`. This way, the include lines will read like `#include <ripple/test/Logic_test.cpp>` for example, but we still get the directory separation.

vinniefalco commented:
Is this rebased against *0.33.0-b7* ?

bachase commented:
Yes, should be 0.33.0-b7 looking at my local gitlog.  

I'll ponder the alternate directory strucutre.  Moving things like that is enough if it makes more sense.

bachase commented:
Is there a way to kick appveyor and travis again?  Appveyor appears to have timed out grabbing a dependency.  travis timed out uploading code coverage results.

Inspecting travis buils manually:
Old (0.33.0-b7)
 117 suites, 485 cases
   lines......: 71.8% (50294 of 70038 lines)
  functions..: 81.4% (15567 of 19116 functions)
New 
 118 suites, 483 cases
  lines......: 64.9% (33980 of 52355 lines)
  functions..: 79.9% (13270 of 16617 functions)

vinniefalco commented:
I kicked them both

bachase commented:
Looks like the discrepancy in number of test cases is due to RIPPLE_ROCKSDB_AVAILABLE no longer being defined in Backend_test.cpp.  Will track it down later and push one more update.  Maybe CI will be happy this time.

bachase commented:
> I just had a really uhh... stupid/smart idea. We can put the test files in src/test/ripple/test. And then set the added C++ include path to src/test. This way, the include lines will read like #include <ripple/test/Logic_test.cpp> for example, but we still get the directory separation. I'm not saying to do that. But its something to think about.

Since `#include <test/Logic_test.cpp>` and similar are only in the `test_unity*.cpp` files, I don't see a big advantage of supporting `#include <ripple/test/Logic_test.cpp>` instead.  Although, I admit seeing ripple in that path is nice to help mirror how things are organized in `src`.

mellery451 commented on src/test/SemanticVersion_test.cpp:
```diff
@@ -0,0 +1,260 @@
+//------------------------------------------------------------------------------
```
I think this (and all other new files) are using tabs for indent. I think CodingStyle.md says we use 4 spaces, no tabs.

bachase replied:
Argh. I'll blame visual studio but will fix. Good catch.

mellery451 commented:
👍

vinniefalco commented:
Files modified:
```
M       .gitignore
M       Builds/VisualStudio2015/RippleD.vcxproj
M       Builds/VisualStudio2015/RippleD.vcxproj.filters
M       CMakeLists.txt
M       SConstruct
D       src/ripple/app/tests/AccountTxPaging.test.cpp
D       src/ripple/app/tests/AmendmentTable.test.cpp
D       src/ripple/app/tests/CrossingLimits_test.cpp
D       src/ripple/app/tests/DeliverMin.test.cpp
D       src/ripple/app/tests/Flow_test.cpp
D       src/ripple/app/tests/HashRouter_test.cpp
D       src/ripple/app/tests/LoadFeeTrack.test.cpp
D       src/ripple/app/tests/MultiSign.test.cpp
D       src/ripple/app/tests/Offer.test.cpp
D       src/ripple/app/tests/OfferStream.test.cpp
D       src/ripple/app/tests/OversizeMeta_test.cpp
D       src/ripple/app/tests/Path_test.cpp
D       src/ripple/app/tests/PayChan_test.cpp
D       src/ripple/app/tests/Regression_test.cpp
D       src/ripple/app/tests/SHAMapStore_test.cpp
D       src/ripple/app/tests/SetAuth_test.cpp
D       src/ripple/app/tests/SetRegularKey.test.cpp
D       src/ripple/app/tests/SusPay_test.cpp
D       src/ripple/app/tests/Taker.test.cpp
D       src/ripple/app/tests/Transaction_ordering_test.cpp
D       src/ripple/app/tests/TxQ_test.cpp
D       src/ripple/app/tests/ValidatorList_test.cpp
D       src/ripple/basics/tests/CheckLibraryVersions.test.cpp
D       src/ripple/basics/tests/KeyCache.test.cpp
D       src/ripple/basics/tests/RangeSet.test.cpp
D       src/ripple/basics/tests/StringUtilities.test.cpp
D       src/ripple/basics/tests/TaggedCache.test.cpp
D       src/ripple/basics/tests/base_uint.test.cpp
D       src/ripple/basics/tests/contract.test.cpp
D       src/ripple/basics/tests/hardened_hash_test.cpp
D       src/ripple/basics/tests/mulDiv.test.cpp
D       src/ripple/beast/clock/tests/beast_abstract_clock_test.cpp
D       src/ripple/beast/clock/tests/beast_basic_seconds_clock_test.cpp
D       src/ripple/beast/container/tests/aged_associative_container.test.cpp
D       src/ripple/beast/core/LexicalCast.cpp
M       src/ripple/beast/core/SemanticVersion.cpp
M       src/ripple/beast/core/core.unity.cpp
D       src/ripple/beast/define_print.cpp
D       src/ripple/beast/hash/impl/hash_speed_test.cpp
D       src/ripple/beast/hash/tests/hash_append_test.cpp
D       src/ripple/beast/hash/tests/hash_speed_test.cpp
M       src/ripple/beast/net/detail/Parse.h
M       src/ripple/beast/net/impl/IPAddressV4.cpp
D       src/ripple/beast/net/tests/IPEndpoint.test.cpp
D       src/ripple/beast/net/tests/beast_asio_error_test.cpp
D       src/ripple/beast/nudb/test/beast_nudb_callgrind_test.cpp
D       src/ripple/beast/nudb/test/beast_nudb_recover_test.cpp
D       src/ripple/beast/nudb/test/beast_nudb_store_test.cpp
D       src/ripple/beast/nudb/test/beast_nudb_verify_test.cpp
D       src/ripple/beast/unity/beast_clock_unity.cpp
D       src/ripple/beast/unity/beast_container_unity.cpp
M       src/ripple/beast/unity/beast_hash_unity.cpp
M       src/ripple/beast/unity/beast_net_unity.cpp
D       src/ripple/beast/unity/beast_nudb_unity.cpp
M       src/ripple/beast/unity/beast_utility_unity.cpp
M       src/ripple/beast/utility/src/beast_Debug.cpp
M       src/ripple/beast/utility/src/beast_PropertyStream.cpp
D       src/ripple/beast/utility/test/beast_Journal_test.cpp
D       src/ripple/beast/utility/test/beast_Zero_test.cpp
D       src/ripple/beast/utility/test/beast_tagged_integer_test.cpp
D       src/ripple/beast/utility/test/beast_weak_fn_test.cpp
M       src/ripple/core/impl/Workers.cpp
D       src/ripple/core/tests/Config.test.cpp
D       src/ripple/core/tests/Coroutine.test.cpp
D       src/ripple/core/tests/SociDB.test.cpp
D       src/ripple/core/tests/Stoppable.test.cpp
D       src/ripple/json/tests/Object.test.cpp
D       src/ripple/json/tests/Output.test.cpp
D       src/ripple/json/tests/Writer.test.cpp
D       src/ripple/json/tests/json_value.test.cpp
D       src/ripple/ledger/tests/BookDirs_test.cpp
D       src/ripple/ledger/tests/Directory_test.cpp
D       src/ripple/ledger/tests/PaymentSandbox_test.cpp
D       src/ripple/ledger/tests/PendingSaves_test.cpp
D       src/ripple/ledger/tests/SkipList_test.cpp
D       src/ripple/ledger/tests/View_test.cpp
D       src/ripple/nodestore/tests/Backend.test.cpp
D       src/ripple/nodestore/tests/Basics.test.cpp
D       src/ripple/nodestore/tests/Database.test.cpp
D       src/ripple/nodestore/tests/Timing.test.cpp
D       src/ripple/nodestore/tests/import_test.cpp
D       src/ripple/nodestore/tests/varint_test.cpp
D       src/ripple/overlay/tests/TMHello.test.cpp
D       src/ripple/overlay/tests/cluster_test.cpp
D       src/ripple/overlay/tests/manifest_test.cpp
D       src/ripple/overlay/tests/short_read.test.cpp
D       src/ripple/peerfinder/tests/Livecache.test.cpp
D       src/ripple/peerfinder/tests/PeerFinder_test.cpp
D       src/ripple/protocol/tests/BuildInfo.test.cpp
D       src/ripple/protocol/tests/IOUAmount.test.cpp
D       src/ripple/protocol/tests/InnerObjectFormats.test.cpp
D       src/ripple/protocol/tests/Issue.test.cpp
D       src/ripple/protocol/tests/PublicKey_test.cpp
D       src/ripple/protocol/tests/Quality.test.cpp
D       src/ripple/protocol/tests/STAccount.test.cpp
D       src/ripple/protocol/tests/STAmount.test.cpp
D       src/ripple/protocol/tests/STObject.test.cpp
D       src/ripple/protocol/tests/STTx.test.cpp
D       src/ripple/protocol/tests/SecretKey_test.cpp
D       src/ripple/protocol/tests/Seed_test.cpp
D       src/ripple/protocol/tests/XRPAmount.test.cpp
D       src/ripple/protocol/tests/digest_test.cpp
D       src/ripple/protocol/tests/types_test.cpp
M       src/ripple/resource/impl/Entry.h
D       src/ripple/resource/tests/Logic.test.cpp
D       src/ripple/rpc/tests/AccountInfo_test.cpp
D       src/ripple/rpc/tests/AccountLinesRPC.test.cpp
D       src/ripple/rpc/tests/AccountObjects_test.cpp
D       src/ripple/rpc/tests/AccountOffers.test.cpp
D       src/ripple/rpc/tests/AccountSet.test.cpp
D       src/ripple/rpc/tests/Book.test.cpp
D       src/ripple/rpc/tests/GatewayBalances.test.cpp
D       src/ripple/rpc/tests/JSONRPC.test.cpp
D       src/ripple/rpc/tests/KeyGeneration.test.cpp
D       src/ripple/rpc/tests/LedgerData.test.cpp
D       src/ripple/rpc/tests/LedgerRequestRPC.test.cpp
D       src/ripple/rpc/tests/RobustTransaction.test.cpp
D       src/ripple/rpc/tests/ServerInfo.test.cpp
D       src/ripple/rpc/tests/Status.test.cpp
D       src/ripple/rpc/tests/Subscribe.test.cpp
D       src/ripple/server/tests/Server_test.cpp
D       src/ripple/shamap/tests/FetchPack.test.cpp
D       src/ripple/shamap/tests/SHAMap.test.cpp
D       src/ripple/shamap/tests/SHAMapSync.test.cpp
D       src/ripple/test/impl/BasicNetwork_test.cpp
D       src/ripple/test/impl/WSClient_test.cpp
D       src/ripple/test/jtx/impl/Env_test.cpp
D       src/ripple/unity/app_tests.cpp
M       src/ripple/unity/basics.cpp
M       src/ripple/unity/beast.cpp
M       src/ripple/unity/core.cpp
M       src/ripple/unity/json.cpp
M       src/ripple/unity/ledger.cpp
M       src/ripple/unity/nodestore.cpp
M       src/ripple/unity/overlay.cpp
M       src/ripple/unity/peerfinder.cpp
M       src/ripple/unity/protocol.cpp
M       src/ripple/unity/resource.cpp
M       src/ripple/unity/rpcx.cpp
M       src/ripple/unity/server.cpp
M       src/ripple/unity/shamap.cpp
M       src/ripple/unity/soci_ripple.cpp
M       src/ripple/unity/test.cpp
```

vinniefalco commented:
There's a lot of translation units in one folder and its hard to browse. We should bring some of the hierarchy back:
```
test/app/
test/basics/
test/beast/
test/core/
test/json/
test/ledger/
test/nodestore/
test/overlay/
test/peerfinder/
test/protocol/
test/resource/
test/rpc/
test/server/
test/shamap/

// These loose files can go in test/ and we
// can figure out what to do with them later:
src/ripple/test/impl/BasicNetwork_test.cpp
src/ripple/test/impl/WSClient_test.cpp
src/ripple/test/jtx/impl/Env_test.cpp
```

See previous GitHub comment for the original list of modified and deleted files to know what goes where.

bachase commented:
Will do. I can make corresponding unity sources for those took rather than the arbitrary numbering.

vinniefalco commented:
test/app is going to be over-represented I think, you might have to break that up into a couple of unity files (not sure though).

Also, not for this PR but the code in `src/ripple/test` is all test code (the **jtx** framework in particular) and will have to be moved but its not quite a simple rename.

bachase commented:
Updated the directory structure.  I also added a call to `lcov --remove src/ripple/test` to ignore that from code coverage for now.  I suppose the remove approach could've been done instead of moving all the files around, but it seems worth having the tests out of the `src/ripple` tree anyway.

AppVeyor died downloading dependencies again =(

vinniefalco commented:
@bachase Over time I want to migrate the structure of the repository to resemble Boost with test separate, doc directory, etc.. and hopefully parts of rippled broken up into individual libraries that are stand-alone (Ed's work on this is promising).

vinniefalco commented:
Link errors building debug/unity
```
1>app_main.obj : error LNK2019: unresolved external symbol "public: void __cdecl ripple::LedgerMaster::setMinValidations(int,bool)" (?setMinValidations@LedgerMaster@ripple@@QEAAXH_N@Z) referenced in function "public: virtual bool __cdecl ripple::ApplicationImp::setup(void)" (?setup@ApplicationImp@ripple@@UEAA_NXZ)
1>app_misc.obj : error LNK2019: unresolved external symbol "public: int __cdecl ripple::LedgerMaster::getMinValidations(void)" (?getMinValidations@LedgerMaster@ripple@@QEAAHXZ) referenced in function "public: virtual class Json::Value __cdecl ripple::NetworkOPsImp::getServerInfo(bool,bool)" (?getServerInfo@NetworkOPsImp@ripple@@UEAA?AVValue@Json@@_N0@Z)
1>
```

bachase commented:
Hrmm, my VS debug unity from scratch just went through ok.  I ran scons again but don't see any diffs in the project files...

vinniefalco commented:
I suspect one of the unity.cpp files is missing an include. Find the translation unit that has the missing symbol, and try to convince yourself that it is being included in the unity build.

bachase commented:
Yes, that is what I tried,  `setMinValidations` is defined https://github.com/ripple/rippled/blob/1ee6ebcf6251c2a4f847e59276de2de9a74406e7/src/ripple/app/ledger/impl/LedgerMaster.cpp#L1323.

I see `LedgerMaster.cpp` in https://github.com/ripple/rippled/blob/1ee6ebcf6251c2a4f847e59276de2de9a74406e7/src/ripple/unity/app_ledger.cpp#L40

And I see `app_ledger.cpp` in 
https://github.com/ripple/rippled/blob/1ee6ebcf6251c2a4f847e59276de2de9a74406e7/Builds/VisualStudio2015/RippleD.vcxproj#L3271

vinniefalco commented:
Hmm... goofing around with app_ledger.cpp and then reverting it made the problem disappear for me. I'm willing to chalk it up to Microsoft suck.

vinniefalco commented:
Great job
:+1:

vinniefalco commented:
Part of #1831 now
